### ISSUE TITLE
nul (0x00) characters are converted to newline or space characters when written to a register

### ISSUE THREAD

**Issue by Rich-YoYo** [vim/vim#12027]
=== Problem

It seems to be  impossible to set a register to a string that contains non-printable ASCII characters; such characters are converted to newlines or spaces depending on the method. For example, attempting to set a register to a single nul (0x00) character with 'setreg'r', nr2char(0))' results in a newline character being copied to the register

I originally thought this problem was confined to nul characters, but it seems to affect any non-printable character

### Steps to reproduce

# Example 1

" Set register 'r' to a null (0x00) character
call setreg('r', nr2char(0))

" Paste register 'r'
"rp

" Error - instead of pasting 0x00, it pasts a newline - examination of the registers shows that register 'r' is set to a newline and not a nul (0x00)

# Example 2

Given a buffer containing a nul (0x00) character, place the cursor over it and yank it
v"ry

" Paste register 't'
"rp

" Error - instead of pasting 0x00, it pasts an ASCII space (0x20) - examination of the registers shows that register 'r' is set to a space and not a nul (0x00)

### Expected behaviour

I expect non-printable characters NOT to be converted to newline/space characters when written to a register

### Version of Vim

Vim 9.0, patches 1-1321

### Environment

OS: Linux
Terminal: Terminalogy
TERM = xterm
Shell: bash

**Comment by zeertzjq** [comment]
- Example 1: Cannot reproduce. And `nr2char(0)` is an empty string.
- Example 2: Cannot reproduce either.

**Comment by tonymec** [comment]
I think this is intentional and documented: see :help NL-used-for-Nul

Best regards,
Tony.

**Comment by Rich-YoYo** [comment]
@tonymec

:help NL-used-for-Nul says that <nul> is only converted to <cr> if fileformat is mac (which it isn't in my case).
While I accept that this is documented (for a fileformat of mac), even then it seems to be very arbitrary behaviour - why would anyone want this?

This also doesn't explain why <cr> gets converted to ASCII 0x20 (space) in some cases.

**Comment by Rich-YoYo** [comment]
@zeetzjq

I thought it was an empty string too, but if I paste it, I get a space (ASCII 0x20) character.

**Comment by chrisbra** [comment]
> :help NL-used-for-Nul says that 'nul' is only converted to 'cr' if fileformat is mac 

That is below `:h CR-used-for-NUL`. 

The thing is, Since `\0` is used as string terminator inside C chars, we cannot get a `\0` inside any register, so we have to convert it. That is what you see here.

**Comment by Rich-YoYo** [comment]
@chrisbra
I hear what you are saying but...

I would expect (and I don't know - I've not checked the code) but I would expect the implementation of the registers to be data/len rather than relying on a nul-terminated string - but I don't know

If I copy nul nul (two characters) to the global clipboard then paste them in vim, I get a space (0x20) and a newline (0x0a) so that's not even consistent behaviour - it's not treating the nul chars equally. It doesn't modify the clipboard (I wouldn't expect it to) but it's clearly able to read it otherwise it wouldn't even get as far as it does

nvim (sorry to mention them :-)) _mostly_ doesn't have this problem (nvim does have a problem but, rather oddly, only when a single nul character is provided in a call to setreg() - it's fine if you give to two nuls or visually yank a single nul) 

As it stands, yank and paste are completely broken in vim if the text contains any nul characters

**Comment by chrisbra** [comment]
> As it stands, yank and paste are completely broken in vim if the text contains any nul characters

Text Files and Ascii NUL is a bit contradicting. a lot of general purpose text processing utilities will not be able to handle that correctly. So, what kind of data are you editing?

**Comment by Rich-YoYo** [comment]
I've written a hex editor using the xxd utility. Obviously this is to edit binary files.

I completely accept that Vim is not really designed to deal with non-plaintext files, but it generally manages it - apart from this aspect with yank and paste (and the 'regsister' functionality that they use)

**Comment by brammool** [comment]
> ### Steps to reproduce
> 
> " Set register 'r' to a null (0x00) character
> call setreg('r', nr2char(0))
> 
> " Paste register 'r'
> "rp
> 
> " Error - instead of pasting 0x00, it pasts a newline

I cannot reproduce this, it puts nothing.  

> ### Expected behaviour
> 
> Given a register containing one or more nul (0x00) characters, I
> expect pasting it to a buffer to paste nul (0x00) chracters
> 
> Instead, the nul characters are pasted as newlines (0x0a)
> 
> I originally thought this problem was confined to nul characters, but
> it seems to affect any non-printable character

Please give a specific example for this.

From the edited message (which is not reflected in the email):

> Given a buffer containing a nul (0x00) character, place the cursor
> over it and yank it
> v"ry
> 
> " Paste register 't'
> "rp
> 
> " Error - instead of pasting 0x00, it pasts an ASCII space (0x20) -
> examination of the registers shows that register 'r' is set to a space
> and not a nul (0x00)

For me this works as expected, the NUL character is pasted.  I inserted
the NUL character using CTRL-V x 00 in Insert mode.

To make sure your examples are reproducible, please start Vim with the
"--clean" argument.  Also try using a commonly used terminal, I have
never seen "Terminalogy".

-- 
hundred-and-one symptoms of being an internet addict:
158. You get a tuner card so you can watch TV while surfing.

 /// Bram Moolenaar -- ***@***.*** -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by Rich-YoYo** [comment]
Using xterm....

> vim --clean

1/ call setreg('r', '^@')
...results in reg 'r' containing ^J (newline) and that is what I get when I paste it

2/ call setreg('r', 'nr2char(0))
...results in reg 'r' containing nothing at all but looking at :help nr2char() explains why this would be

3/ call setreg('r', '^@^A^B')
...results in reg 'r' containing ^J^A^B

4/ Setting the buffer to

   ^@^A^B

...and visually yanking the 3 bytes into reg 'r' results in reg 'r' containing "^@^A^B" which is exactly what I would hope for
Pasting reg 'r' into a buffer also results in ^@^A^B - excellent!

-----
I also tried without --clean and the results were the same

-----
So it's different behaviour to what I originally posted but still not consistent; yank behaves differently to setreg().
My results also seem to be inconsistent with what others have reported, which is a mystery to me

Also, contrary to what I observed earlier, it seems that non-printable characters other than nul (0x00) do indeed seem to be preserved - I can offer no explanation as to why I saw something different previously

I went back to my normal terminal/environment and got the same results as above except that for test 4, 'r' gets set to ^J^A^B. So there is definitely an issue created by the terminal but I'm at a loss as to how the terminal type could effect this functionality (but it's not something I have much experience with)

So in summary, I still think the behaviour is not 'correct' (consistent), but its behaving (for me, at least) very confusingly

And apologies for what now seem to be red herrings that I previously posted

**Comment by brammool** [comment]
> Using xterm....
> 
> > vim --clean
> 
> 1/ call setreg('r', '^@')
> ...results in reg 'r' containing ^J (newline) and that is what I get when I paste it

When I read this I see '^@' as a string with two characters, '^' and
'@'.  I suppose you actually have something else.  How did you type
this?


-- 
Keep America beautiful.  Swallow your beer cans.

 /// Bram Moolenaar -- ***@***.*** -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by Rich-YoYo** [comment]
I type it as <CTRL-V>u0000 but I figured that would probably get lost in translation somewhere between my keyboard and my web browser, hence I wrote it as I did
